### ISSUE TITLE
Add +383 country calling code for Kosovo

### ISSUE THREAD

**Issue by kedar2807** [googlei18n/libphonenumber#1486]
I passed number to parse in this format : 383.xx.xxx.xxx(countrycode.areacode...), country iso code : XK
But it does not parse that number.

**Comment by kedar2807** [comment]
Error type: INVALID_COUNTRY_CODE. Missing or invalid default region.

**Comment by kedar2807** [comment]
CountryCodeToRegionCodeMap class does not contain this country code mapping.

**Comment by kkeshava** [comment]
@kedar2807,

Thank you for taking the time to improve libphonenumber!

As of now Kosovo shares mobile numbers with Monaco (MC, +377) and Slovenia (SI, +386). We have not added separate metadata entry for Kosovo. We need official evidence either from ITU (e.g. https://www.itu.int/dms_pub/itu-t/oth/02/02/T020200008D0004PDFE.pdf) or from national numbering authority of Kosovo to add the country in our library.

Please share any official evidence you may have. See our [guidelines](https://github.com/googlei18n/libphonenumber/blob/master/CONTRIBUTING.md#provide-supporting-evidence).

**Comment by kedar2807** [comment]
Here's the official document : 
http://www.itu.int/dms_pub/itu-t/opb/sp/T-SP-E.164D-2016-PDF-E.pdf

**Comment by kkeshava** [comment]
@kedar2807,

The document provided by you only gives the list of assigned country codes. It does not have the numbering plan for Kosovo (example provided earlier) which lists all the phone numbers like fixedline, mobile etc. We need official evidence which lists all the categories to add it in our library.

**Comment by keghani** [comment]
*   In our metadata [0]:
    *   Kosovo shares country calling code +377 with Monaco, where 044 and 045 are mobile in Kosovo.
    *   Kosovo shares country calling code +381 with Serbia, where 028, 029, 038, and 039 are fixed-line in Kosovo.
    *   Kosovo shares country calling code +386 with Slovenia, where 043 and 049 are mobile in Kosovo.
*   +383 is assigned to Kosovo. [1], [2]
*   **The Kosovo ranges under +377, +381, and +386 seem to still be in use.**
*   **It's unclear which ranges, if any, are assigned under +383 at this time.**

Due to those last 2 points we are still missing information.
In order to add +383 as a country calling code in the library, we need to know which ranges are valid under the new plan, their start dates, and possible transitions from the old ranges.
I think that the ITU will update this in due time.

For the time being, @kedar2807 are you aware of any numbers under +383 that are actually in use?


[0] https://raw.githubusercontent.com/googlei18n/libphonenumber/16d8f47ec1602d5a5eee6187e90ceea0a556dcba/resources/PhoneNumberMetadata.xml
[1] https://en.wikipedia.org/wiki/Telephone_numbers_in_Kosovo
[2] https://www.itu.int/dms_pub/itu-t/opb/sp/T-SP-E.164D-2016-PDF-E.pdf

**Comment by keghani** [comment]
Since the ranges are unclear at this time, I'll close this issue, and the library's working as intended.
We're subscribed to ITU notifications so if they release a doc we should know about it.

Still, feel free to re-open when this becomes an issue.

**Comment by avimar** [comment]
There's news, but no official ITU documentation:

https://www.telegeography.com/products/commsupdate/articles/2017/02/06/kosovo-begins-using-new-dialling-code/

**Comment by andyst** [comment]
Thanks for the update, @avimar.

If you (or someone else on this thread) would like to accelerate support of +383 numbers in libphonenumber, you can help by contacting the right people to get the numbering plans published through the right channels.

## ITU
E.g. contacting ARKEP (maybe other organizations too?) to get them to publish the numbering plan through the ITU.

This would result in Kosovo / +383 to be listed under https://www.itu.int/oth/T0202.aspx?parent=T0202 (there's also more information there on how to file / publish numbering plans). We actively monitor ITU updates and use these numbering plans as a basis for the metadata in libphonenumber.

If this doesn't happen soon, it'd be useful to understand when the numbering plan will be published through the ITU (and when the first +383 numbers will be operational for end-users).

## Mobile Network Operators & GSMA
Please also contact mobile network operators to publish their assigned numbers / numbering plan as IR.21 document with the GSMA. Publishing (mobile) numbering plans in a timely manner with the GSMA ensures that customers don't have problems with products / services which depend on libphonenumber (such as Google, Facebook, and many others).

## Example Numbers
If you know of a specific +383 number that we could call to verify that it connects / that it's operational, please let us know. You can send such numbers also to libphonenumber-supporting-docs@googlegroups.com if you'd like to only share it with the libphonenumber team. Please specify the type (fixed-line, mobile, ...) of the phone number. We prefer an official example phone number that can be called (some countries do have that). A number by a business (e.g. with a website as evidence) would also be great.

Thanks!

**Comment by rvandam** [comment]
Would this example be sufficient: https://www.facebook.com/pg/GlobalPlusAutomotive/about/ ? Lists the number as +383 38 731 595

**Comment by andyst** [comment]
1.  We tried to dial +383 38 731 595 from India and from Switzerland (mobile and fixed-line) and could not connect successfully. This could very well be a local problem with the carriers / operators in Switzerland / India, or something between them and the carrier / operator of this number. But it's not working yet. I'm excited though that we might be very close to witnessing the first international phone calls to +383 / the official country calling code for Kosovo.

2.  It is not sufficient to have an example for us to add support for +383. Please see my comment from Feb 8.

Thanks!

**Comment by andyst** [comment]
I'm closing as this issue is infeasible for us to fix at this point (we need
more information). Please re-open if you respond so we pick up on it.

Please see my comment from Feb 8 on what's needed to unblock this issue.

Thanks!

**Comment by andyst** [comment]
**Does anyone know of a +383 number which we could call from abroad** (e.g. Switzerland or India)?

--

The good news is that there's an ITU document for Kosovo now which lists fixed-line and mobile number ranges:
https://www.itu.int/dms_pub/itu-t/oth/02/02/T02020000FD0001PDFE.pdf

Unfortunately the contact information in there uses a Serbian phone number (+381). We'd like to make a test phone call to a +383 number if possible.

Thanks!

**Comment by angeloprobst** [comment]
Now Kosovo is officially documented by ITU: https://www.itu.int/dms_pub/itu-t/oth/02/02/T02020000FD0001PDFE.pdf
We are waiting for this support.
Thank you!

**Comment by andyst** [comment]
That's not new. See my comment from June 12th above, we still need to verify that +383 numbers can be called from abroad before we can add this country calling code to libphonenumber's metadata.

-> **Does anyone know of a +383 number which we could call from abroad** (e.g. Switzerland or India)?

**Comment by ckromo** [comment]
ARKEP has announced that more than 160 operators worldwide can now reach Kosovo on +383 (I think they were referring to Vala/Kosovo Telecom). They have said all fixed-line phone numbers will be switched to new code and the old one (381) will be switched off by the years end.

You can try to rech ARKEP themselves on +383 (0) 38 212 345 (http://arkep-rks.org/?cid=1,3).

Or, here you have a business that lists their land-line and mobile phones numbers with the new code: http://www.bakiautomobile.com/kontakti/
+383 (0) 38 600 041 (Kosovo Telecom - landline); +383 (0) 44 505 499 (Vala/Kosovo Telecom - mobile).

**Comment by caktor** [comment]
ARKEP announced recently that more than 160 operators worldwide can now reach Kosovo on +383 (I think they were referring to Vala/Kosovo Telecom). They have said previously that all fixed-line phone numbers will be switched to the new code and the old one (381) will be terminated by the years end.

It's almost years end so you can try to confirm this by reaching ARKEP themselves on +383 38 212 345 (http://arkep-rks.org/?cid=1,3).

Or, here you have a business that lists their landline and mobile phone numbers with the new code: http://www.bakiautomobile.com/kontakti/
+383 38 600 041 (Kosovo Telecom - landline); +383 44 505 499 (Vala/Kosovo Telecom - mobile).

If you are successful on reaching both mobile and landlines and want to go ahead on implementing the changes I can then supply you with most recent updates on new allocations since the last ITU communicaton (linked on above comments).

**Comment by keghani** [comment]
That's interesting, thanks for sharing @caktor!

So those numbers do seem to connect.

I see an announcement at http://arkep-rks.org/?cid=2,23,1002
According to this, the first phase's conclusion is June 2018. Do you have another source saying that the deadline is end of 2017?

Yes, we'd be very interested in an updated list of allocations. Please see our [contribution guidelines](https://github.com/googlei18n/libphonenumber/blob/master/CONTRIBUTING.md) for evidence requirements, also how to share in case the document is private.

**Comment by caktor** [comment]
I'm sorry, I was trying to find any recent written announcement from ARKEP about their progress but could not find any. Government agencies, not known for good PR after all. That announcement in English was not available in Albanian and was clearly meant for and sent to international operators. But all other decisions are available in Albanian.

I also tried to find articles in English but could not find any recent one regarding the progress on implementation of the new code. Nevertheless I will post here some sources in Albanian. I can help on translation, should you need it.


**November 10th article**
http://www.koha.net/arberi/56521/kodi-383-ne-proces-te-operacionalizimit/

In this article Chairman of ARKEP states that beginning 2018 all old foreign codes will start phasing out. That will end on June 30th 2018 when only +383 will be active.

Also, in the same article, Technical Director of Kosovo Telecom regarding the use of +381 code on fixed-line numbers says that they are already ready to switch if off. They're just waiting a formal decision from ARKEP. Kosovo Telecom has the absolute majority of the market on landlines.

**November 2nd**
https://vimeo.com/241006371

This is a video recording of the Chairman of ARKEP reporting to the Parliamentary Committee for Economic Development. On 6:14 onwards he says that +383 on mobile operations will be fully implemented until June 2018, but regarding landline numbers that will happen until December 31st 2017.

I will, in the next comment, provide numbering plan and range allocations based on ARKEP decisions.

**Comment by caktor** [comment]
First off, I will contact ARKEP and ask them to publish all documents in English too. But for the time being I will link to the ones in Albanian and will translate those for MNC, fixed-line and mobile ranges.

**List of MNC codes allocated to mobile operators**
http://www.arkep-rks.org/repository/docs/Lista%20e%20kodeve%20MNC%20t%C3%AB%20caktuara%20nga%20ARKEP%202017.pdf

MCC | MNC | Operator
----- | ------| ---------
221 | 01 | Vala (Telekomi i Kosovës sh.a.)
221 | 02 | IPKO (IPKO Telecommunications LLC)
221 | 06 | Z-Mobile (Dardafon.Net LLC) - MVNO on Vala network



**List of allocated ranges of geografic numbers (fixed-line)**
http://arkep-rks.org/repository/docs/Numeracioni%20Gjeografik%20-%20Telefonia%20Fikse%20(Seria%202x%20dhe%203x).pdf

Geographic numbering on landlines, as of now, is regulated as follows; formating as on the above link:

Series: 2x and 3x
NDC: 2-3 digits SN: 5-6 digits
Formats: AB-CDEFGH, ABC-DEFGHI, ABC-DEFGH

Area code (NDC) | Region
----------------- | -------
28 | Mitrovica
280 | Gjilan
29 | Prizren
290 | Ferizaj
38 | Prishtina
39 | Peja
390 | Gjakova



**List of non-geographic numbers (mobile)**
http://arkep-rks.org/repository/docs/Numeracioni%20Jo-Gjeografik%20mobil%20(Seria%204x).pdf

Non-geographic numbering scheme is regulated as follows:

Series: 4x
NDC: 2 digits, SN: 6 digits
Format: AB-CDEFGH

Code | Block Start | Block end | Operator
----- | ----------- | ---------- | ---------
43 | 100000 | 199999 | IPKO [Reserved - not in use]
43 | 200000 | 499999 | D3 Mobile
43 | 500000 | 999999 | IPKO
44 | 000000 | 999999 | Vala
45 | 100000 | 499999 | Vala
45 | 500000 | 699999 | Z-Mobile
45 | 700000 | 999999 | Vala
46 | 100000 | 199999 | Z-Mobile
47 | 100000 | 159999 | mts
49 | 000000 | 099999 | IPKO [Reserved - not in use]
49 | 100000 | 999999 | IPKO


This is the National Numbering Plan (in Albanian):
http://arkep-rks.org/repository/docs/Reg%20Nr%2032%20-%20Plani%20Kombetar%20i%20Numeracionit%20per%20Rrjetat%20e%20Komunikimeve%20Elektronike.pdf

And here you can find all other series and ranges:
http://arkep-rks.org/?cid=1,50

**Comment by keghani** [comment]
@caktor,

Thanks for the links, they're very valuable!

It looks like +383 numbers are starting to connect, and now we're looking into adding support for them with one of the first releases of the new year.

I noticed that your comments are getting deleted from the GitHub page. Would you mind if future comments stay? It's hard to refer back to everything when they're deleted. Alternatively you can email us at an address that only core team members have access to, the setup instructions are [here](https://github.com/googlei18n/libphonenumber/blob/master/CONTRIBUTING.md#provide-authoritative-evidence).

Please note that you have submitted this information with the knowledge that we
may use, modify, reproduce, publicly display, and distribute the information in
accordance with our open source policies. Please let us know if you did not
intend to submit the information for such use.

**Comment by komiku** [comment]
@keghani yeah a mishap with GitHub. Apparently GitHub suspends accounts with disposable emails (not a software developer; didn't want my email to be tied to yet another service). Nevertheless, I'm sucked to this now and will try to contribute regularly.

The links were all public documents and you can use them freely. I will post them again for reference. I've also contacted ARKEP to have numbering plan and all future decisions published in English too and asked for some clarification regarding the progress.

I will make the changes for carrier and geocoding and will try to edit the other 3 metadata files. I don't suppose there are other files to change to add support, are there? You can maybe review and merge them whenever you decide.

**Comment by komiku** [comment]
Geographic numbers
---------
**List of allocated ranges of geografic numbers (fixed-line)**
http://arkep-rks.org/repository/docs/Numeracioni%20Gjeografik%20-%20Telefonia%20Fikse%20(Seria%202x%20dhe%203x).pdf

Series: 2x and 3x
NDC: 2-3 digits SN: 5-6 digits
Formats: AB-CDEFGH, ABC-DEFGHI, ABC-DEFGH

Area code (NDC) | Region
----------------- | -------
28 | Mitrovica
280 | Gjilan
29 | Prizren
290 | Ferizaj
38 | Prishtina
39 | Peja
390 | Gjakova





Non-geographic numbers (mobile):
-----

**List of non-geographic numbers**
http://arkep-rks.org/repository/docs/Numeracioni%20Jo-Gjeografik%20mobil%20(Seria%204x).pdf

Series: 4x
NDC: 2 digits, SN: 6 digits
Format: AB-CDEFGH

Code | Block Start | Block end | Operator
----- | ----------- | ---------- | ---------
43 | 100000 | 199999 | IPKO [Reserved - not in use]
43 | 200000 | 499999 | D3 Mobile
43 | 500000 | 999999 | IPKO
44 | 000000 | 999999 | vala
45 | 100000 | 499999 | vala
45 | 500000 | 699999 | Z Mobile
45 | 700000 | 999999 | vala
46 | 100000 | 199999 | Z Mobile
47 | 100000 | 159999 | mts
49 | 000000 | 099999 | IPKO [Reserved - not in use]
49 | 100000 | 999999 | IPKO


**National Numbering Plan (in Albanian):**
http://arkep-rks.org/repository/docs/Reg%20Nr%2032%20-%20Plani%20Kombetar%20i%20Numeracionit%20per%20Rrjetat%20e%20Komunikimeve%20Elektronike.pdf

**Here you can find all other series and ranges:**
http://arkep-rks.org/?cid=1,50

**Comment by keghani** [comment]
Thanks a lot @komiku! That's excellent research, and thanks for the pull request! Someone from the team will review it soon, and we're targeting to merge it before the first release of 2018 along with our metadata changes.

Thanks for offering to help, it's much appreciated! We haven't had luck hearing back from ARKEP, are you already in touch with them? We still had the following questions in case you know or can find out:

*   Is +383 already connecting for **all** numbers in Kosovo?
*   When will the existing +381, +377, and +386 numbers for Kosovo stop connecting?
    *   Will they happen in one go or in phases?
    *   If in phases, which ranges/prefixes will stop when?
*   Are the phone numbers (other than the calling code part) staying the same?
    *   E.g. is it the case for all existing devices that somebody previously with a phone number of +381 123456 now gets the +383 123456 number?
    *   Or is there some other rule for mapping old numbers to new numbers?
*   What does this mean? From [http://www.koha.net/arberi/56521/kodi-383-ne-proces-te-operacionalizimit/](http://www.koha.net/arberi/56521/kodi-383-ne-proces-te-operacionalizimit/):
    > “Deri më tani, nga shumica absolute e vendeve të botës, Kosova mund të thirret ndërmjet kodit +383 nëse përdoret në vend të kodeve ekzistuese, derisa nga fillimi i vitit 2018 do të fillojë një fazë e re ku do të fillohet me largimin e kodeve aktuale dhe faza përfundimtare kur do të largohen këto kode dhe do të përdoret vetëm +383, do të jetë 30 qershori i vitit 2018”, sqaron Gashi.
    *   I understand it overall (from Google translate) but missing the details :)

**Comment by komiku** [comment]
I'm in touch with ARKEP, will update when and if I get updates. About the questions:

- I don't know if +383 is open on all carriers as of now. I would believe that is the case. I don't know if you had the chance to see my first comment before GitHub deleted them but I had posted links to articles with statements from ARKEP and Kosovo Telecom. Basically ARKEP has announced couple of weeks ago that more than 160 carriers can reach numbers in Kosovo through +383. He may have referred to vala/Kosovo Telecom since it's the biggest mobile carrier and state owned. Also the Technical Director of Kosovo Telecom has said that they are ready for and are just waiting for a formal decision from ARKEP to shut off +381 on fixed-line numbers (they hold the absolute majority on landlines).

- +381 is slated to be shut on 31st of December 2017 on landlines. However the recent mobile carrier to be licensed (albeit on a temporary license), mts doo, uses this prefix too. They actually had to issue SIM cards with double numbers, the one users held before from the serbian operators, which ceased operation in Kosovo, and the new one from the range allocated to them from ARKEP. So it's not clear yet if they will have to switch earlier than other mobile carriers. I've asked ARKEP about this, will update.
+377 and +386 are slated to be completely switched off on 30th of June 2018. As that article you have linked says, beginning 2018 a new phase will begin, when the old codes will start to be phased out gradually and 30th of June will be the deadline to cut support (my guess is operators will use traffic data to determine which paths to shut first).

- Phone numbers: NSNs are staying the same at this time. The National Numbering Plan Regulation was on public consultation and ARKEP had proposed to change fixed-line area codes to two digits for all regions. I guess the operators were not ready and now this is left to be done after the new international code implementation is completed first. So for now all the numbers stay the same.

**Comment by andyst** [comment]
@komiku:
Do you have an official doc by ARKEP / an official authority which states "+381 is slated to be shut on 31st of December 2017 on landlines"?

It's important we know the date on which this will happen and companies which store phone numbers (e.g. of businesses or for user authentication) have time to prepare for this event.

Thanks!
Andy

**Comment by komiku** [comment]
@andyst 
https://vimeo.com/241006371
Chairman of ARKEP, 6:25-6:34: "In fixed-line telephony it must be implemented by the end of this year, namely by the 31st of December fixed-line telephony will remove +381 completely and will only use +383.

I'm still waiting for a written response from them.

**Comment by komiku** [comment]
Btw, since we're talking about companies, does anyone here know which organizations use this library or the metadata aggregated here in their products? I'm talking about big-name companies. Does Google use them?

**Comment by andyst** [comment]
Thanks for the link. Indeed, we're looking for a written confirmation.

Who's using libphonenumber?
Many big and small tech companies use libphonenumber. Google is among them. Generally, most companies / projects dealing with phone numbers use libphonenumber / its metadata (or one of its ports or a service which uses libphonenumber), with the big exception of the telecom carriers / operators themselves (some of them use libphonenumber for a few things too though).

**Comment by andyst** [comment]
@komiku 
Do you know whether mobile number portability applies to Kosovo +383 numbers?

This is about whether we should set `mobileNumberPortableRegion` to true or false for +383 / XK.

I skimmed through a Google Translate version of 
http://arkep-rks.org/repository/docs/Reg%20Nr%2032%20-%20Plani%20Kombetar%20i%20Numeracionit%20per%20Rrjetat%20e%20Komunikimeve%20Elektronike.pdf and could not find a statement about this.

Context:
https://en.wikipedia.org/wiki/Mobile_number_portability

**Comment by komiku** [comment]
@andyst 
Yes, number portability applies to both geographic numbers and non-geographic ones. On geographic ones you can't change the region. I'm not aware that any operator has started supporting it yet though or ARKEP enforcing it. The regulation has been passed and is in force:
http://arkep-rks.org/repository/docs/Rregullore%20Nr%2033%20per%20Bartshmerine%20e%20Numrave.pdf

Please refer to the previous NNP. It it mostly the same as the new one. You can compare it with the new one you referred and where you have doubt I will help. Article 22 is about portability:
http://arkep-rks.org/repository/docs/Reg%20Nr%2026%20-%20Regulation%20on%20National%20Numbering%20Plan.pdf

**Comment by andyst** [comment]
Thanks, @komiku !
Article 15 of the first doc you mention seems to imply that mobile number portability may not have been implemented yet (blocked by design & implementation of technical specifications).

We may want to add `mobileNumberPortableRegion=true` in a later change and start without it.

**Comment by komiku** [comment]
You are quite right. I didn't read it through. I'll open an issue when it starts applying.

**Comment by andyst** [comment]
I was trying to reach out to ARKEP and RATEL regarding the timeline of +381 no longer terminating or originating in Kosovo (as that will be a disruptive event for e.g. Google users in Kosovo).

@RATEL: I did not reach anyone at RATEL yet.

@ARKEP: For now I only got a confirmation regarding the June 1st, 2018 deadline for all previous calling codes (+377, +381 and +386). Tomorrow morning, I will try to get a confirmation / estimate for a more specific date (for now I suspect it is not Dec 31st, 2017).

**Comment by komiku** [comment]
I got a response to my inquiries earlier today from the Chief of Electronic Communications at ARKEP. They wouldn't confirm of deny anything and only said what they previously had laid in what they call their 'Migration Plan': 

**- June 1st, 2018 is the latest date after which no call should originate from or terminate to previous codes.**

**- December 31st, 2018 is the latest date when E.212 code (221) should be fully implemented.**

Based on this I think we should add a healthy dose of salt to further statements from them and wait for concrete steps. In January **vala** was supposed to start replacing SIM cards. Lets see if that happens.

Regarding the English versions of documents they said they were working on having them translated and will publish them soon.

**Comment by andyst** [comment]
Thanks a lot, @komiku. I was also on the phone with a representative of ARKEP yesterday and today and confirmed that

*   it's up to operators when exactly they stop terminating 381 calls in Kosovo 
*   381 calls will continue to stop terminating in Kosovo for sure until Feb-1 (a somewhat arbitrary date I asked about, probably longer)
*   plans probably changed after getting feedback from operators (that's why it's not Dec-31 2017 anymore)
*   it's up to operators when exactly they stop terminating 381 calls in Kosovo (as long as that happens until June 2018)

**Comment by komiku** [comment]
It's always up to operators because they will have to make the changes. But the problem is that they (ARKEP) are not being clear and transparent and throwing legalese terms very cleverly, using the conditional **'should'** instead of the compulsory **'will'**. Well trained bureaucrats. It makes one be very sceptic.

**Comment by andyst** [comment]
True, not having specific dates makes this rollout actually much harder for all involved parties.

Thanks for all the contributions here (and carrier / geocoding mappings). We are readying for the rollout in January.

**Comment by andyst** [comment]
+383 landed in v8.8.11.

Thanks @komiku and all for all your contributions to make this happen!

https://github.com/googlei18n/libphonenumber/releases/tag/v8.8.11
